readme.md
![secp256k1 logo](https://raw.githubusercontent.com/pur3miish/universal-ecdsa/master/static/secp256k1.svg)

# Universal ECDSA

[![NPM Package](https://img.shields.io/npm/v/universal-ecdsa.svg)](https://www.npmjs.org/package/universal-ecdsa) [![CI status](https://github.com/pur3miish/universal-ecdsa/workflows/CI/badge.svg)](https://github.com/pur3miish/universal-ecdsa/actions) [![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://github.com/pur3miish/eos-ecc/blob/main/LICENSE)

A [Universal](https://en.wikipedia.org/wiki/Isomorphic_JavaScript) JavaScript [Elliptic Curve Digital Signature Algorithm](https://en.bitcoin.it/wiki/Elliptic_Curve_Digital_Signature_Algorithm) (ECDSA) for the Koblitz secp256k1 curve.

Info

_The Web Assembly binary file `esdsa.json` that is consumed by the JavaScript environments is compiled from C/C++, see [cpp/readme.md](https://github.com/pur3miish/universal-ecdsa/tree/master/cpp)._

- \~28 kB _(minifeied + gzipped) to bundle_

# Setup

```shell
npm i universal-ecdsa
```

# Suport

- Node.js `^12.20.1 || >= 13.2`
- Browser `defaults, no IE 11`

_You will need Node +15 web crypto for passing test._

# API

- [function get_public_key](#function-get_public_key)
- [function sha256](#function-sha256)
- [function sign](#function-sign)

## function get_public_key

Generates a compressed public key for the `secp256k1` curve.

| Parameter     | Type       | Description                  |
| :------------ | :--------- | :--------------------------- |
| `private_key` | Uint8Array | secp256k1 valid private key. |

**Returns:** Uint8Array — Public key.

### Examples

_Ways to `import`._

> ```js
> import { get_public_key } from 'universal-ecdsa'
> ```

_Ways to `require`._

> ```js
> const { get_public_key } = require('universal-ecdsa')
> ```

_Usage `get_public_key`._

> ```js
> const private_key = new Uint8Array([
>   210, 101, 63, 247, 203, 178, 216, 255, 18, 154, 194, 126, 245, 120, 28, 230,
>   139, 37, 88, 196, 26, 116, 175, 31, 45, 220, 166, 53, 203, 238, 240, 125
> ])
>
> get_public_key(private_key).then(console.log) // compressed public key.
> ```
>
> > The logged output was \[2, …, 207].

---

## function sha256

Universal sha256 message digest helper function.

| Parameter | Type       | Description          |
| :-------- | :--------- | :------------------- |
| `data`    | Uint8Array | Binary data to hash. |

**Returns:** Uint8Array — Message digest.

### Examples

_Ways to `import`._

> ```js
> import { sha256 } from 'universal-ecdsa'
> ```

_Ways to `require`._

> ```js
> const { sha256 } = require('universal-ecdsa')
> ```

_Usage `sha256` in node._

> ```js
> const array = Uint8Array.from(
>   Buffer.from('The quick brown fox jumped over the lazy dog')
> )
>
> sha256(array).then(console.log)
> ```
>
> > The logged output is \[215, …, 146 ]

---

## function sign

Generates a digital signature on the secp256k1 Koblitz curve.

| Parameter         | Type      | Description            |
| :---------------- | :-------- | :--------------------- |
| `Arg`             | object    | Argument.              |
| `Arg.private_key` | Uin8Array | secp256k1 private key. |
| `Arg.data`        | Uin8Array | Data to sign.          |

**Returns:** Signature — Digital signature object containing `r` and `s` values.

### Examples

_Ways to `import`._

> ```js
> import { sign } from 'universal-ecdsa'
> ```

_Ways to `require`._

> ```js
> const { sign } = require('universal-ecdsa')
> ```

_Usage `sign`._

> ```js
> const private_key = new Uint8Array([
>   210, 101, 63, 247, 203, 178, 216, 255, 18, 154, 194, 126, 245, 120, 28, 230,
>   139, 37, 88, 196, 26, 116, 175, 31, 45, 220, 166, 53, 203, 238, 240, 125
> ])
>
> const data = Uint8Array.from([104, 101, 108, 108, 111])
> sign({ data, private_key }).then(console.log)
> ```
>
> > The logged output is { r: \[23, …, 89], s: \[111, …, 142] }


changelog.md
# universal-ecdsa changelog

## 1.1.3

- Equality issue for [deterministically_generate_k](cpp/src/ecdsa.cpp) for {r,s} values fixed (start with 0x80).

## 1.1.2

### Patch

- racid bug fix.
- Closes [#2](https://github.com/pur3miish/universal-ecdsa/issues/2).

# 1.1.1

- get_public_key is not producing a valid public key, prepends 3 for odd y coordinates and 2 for x.

# 1.1.0

### Minor

- Signature generation now includes racid.

- Private key size validation, closes #1.

## 1.0.0

### Major

Initial Release


public/sign.js
'use strict'

const init_wasm = require('../private/init_wasm.js')
const sha256_hash = require('./sha256.js')

/**
 * Generates a digital signature on the secp256k1 Koblitz curve.
 * @kind function
 * @name sign
 * @param {object} Arg Argument.
 * @param {Uin8Array} Arg.private_key secp256k1 private key.
 * @param {Uin8Array} Arg.data Data to sign.
 * @returns {Signature} Digital signature object containing `r` and `s` values.
 * @example <caption>Ways to `import`.</caption>
 * ```js
 * import { sign } from 'universal-ecdsa'
 * ```
 * @example <caption>Ways to `require`.</caption>
 * ```js
 * const { sign } = require('universal-ecdsa')
 * ```
 *  @example <caption>Usage `sign`.</caption>
 * ```js
 * const private_key = new Uint8Array([
 *   210, 101, 63, 247, 203, 178, 216, 255, 18, 154, 194, 126, 245, 120, 28, 230,
 *   139, 37, 88, 196, 26, 116, 175, 31, 45, 220, 166, 53, 203, 238, 240, 125
 * ])
 *
 * const data = Uint8Array.from([104, 101, 108, 108, 111])
 * sign({ data, private_key }).then(console.log)
 * ```
 * > The logged output is { r: [23, …, 89], s: [111, …, 142] }
 */
async function sign({ private_key, data }) {
  const { _Z9sign_hashv, _Z7get_ptri, set_bytes, extract_bytes } =
    await init_wasm()
  const hash = await sha256_hash(data)
  set_bytes(_Z7get_ptri(2), private_key)
  set_bytes(_Z7get_ptri(3), hash)
  _Z9sign_hashv()

  const r = extract_bytes(_Z7get_ptri(0), 32)
  const s = extract_bytes(_Z7get_ptri(1), 32)
  const [racid] = extract_bytes(_Z7get_ptri(5), 1)

  return { r, s, racid }
}

module.exports = sign
